PayloadPkg/OsLoader/LoadImage.c

```diff
@@ -55,149 +108,127 @@ GetBootImageFromRawPartition (
-    }
+  Status = MediaGetMediaInfo (BootOption->HwPart, &BlockInfo);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Get media info error - %r\n", Status));
+    return Status;
+  }
 
-    Address =  LogicBlkDev.StartBlock + LbaAddr;
+  //
+  // Read the IAS Header first to get total size of the IAS image.
+  // Make sure to round the Header size to be block aligned in bytes.
+  //
+  BlockSize = BlockInfo.BlockSize;
+  AlignedHeaderBlkCnt = (sizeof (IAS_HEADER) + (BlockSize - 1)) / BlockSize;
+  AlignedHeaderSize = AlignedHeaderBlkCnt * BlockSize;
 
-    Status = MediaReadBlocks (
-               BootOption->HwPart,
-               Address,
-               AlignedHeaderSize,
-               BlockData
-               );
+  BlockData = AllocatePages (EFI_SIZE_TO_PAGES (AlignedHeaderSize));
+  if (BlockData == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
 
-    if (EFI_ERROR (Status)) {
-      DEBUG ((DEBUG_INFO, "Read image error - %r\n", Status));
-      return Status;
-    }
+  Address =  LogicBlkDev.StartBlock + LbaAddr;
 
-    //
-    // Make sure to round the image size to be block aligned in bytes.
-    //
-    ContainerHdr = (CONTAINER_HDR *)BlockData;
-    if (ContainerHdr->Signature == CONTAINER_BOOT_SIGNATURE) {
-      ImageSize = ContainerHdr->DataOffset + ContainerHdr->DataSize;
-    } else if (ContainerHdr->Signature == IAS_MAGIC_PATTERN) {
-      ImageSize = IAS_IMAGE_SIZE ((IAS_HEADER *) BlockData);
-    } else {
-      DEBUG ((DEBUG_INFO, "No valid image header found !\n"));
-      return EFI_LOAD_ERROR;
-    }
+  Status = MediaReadBlocks (
+             BootOption->HwPart,
+             Address,
+             AlignedHeaderSize,
+             BlockData
+             );
 
-    //
-    // Check image size from the image header
-    //
-    if (ImageSize == 0) {
-      return EFI_LOAD_ERROR;
-    }
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Read image error - %r\n", Status));
+    return Status;
+  }
 
-    AlignedImageSize = ((ImageSize % BlockSize) == 0) ? \
-                       ImageSize : \
-                       ((ImageSize / BlockSize) + 1) * BlockSize;
-    if (AlignedImageSize > MAX_IAS_IMAGE_SIZE) {
-      DEBUG ((DEBUG_INFO, "Image is bigger than limitation (0x%x). ImageSize=0x%x\n",
-              MAX_IAS_IMAGE_SIZE, AlignedImageSize));
-      //
-      // Free temporary pages used for image header
-      //
-      FreePages (BlockData, EFI_SIZE_TO_PAGES (AlignedHeaderSize));
-      return EFI_LOAD_ERROR;
-    }
+  //
+  // Make sure to round the image size to be block aligned in bytes.
+  //
+  ContainerHdr = (CONTAINER_HDR *)BlockData;
+  if (ContainerHdr->Signature == CONTAINER_BOOT_SIGNATURE) {
+    ImageSize = ContainerHdr->DataOffset + ContainerHdr->DataSize;
+  } else if (ContainerHdr->Signature == IAS_MAGIC_PATTERN) {
+    ImageSize = IAS_IMAGE_SIZE ((IAS_HEADER *) BlockData);
+  } else {
+    DEBUG ((DEBUG_INFO, "No valid image header found !\n"));
+    return EFI_LOAD_ERROR;

```

Comment:
how do you handle BlockData free ?  
many places just return.  Please enhance error handling.

Reply:
it should be handled in UnloadBootImages() if loaded failure.
This patch doesn't change the original logic.

Reply:
OK